telluric/context.py

```diff
@@ -0,0 +1,157 @@
+"""Telluric thread local context manager."""
+import logging
+import threading
+
+
+class ThreadContext(threading.local):
+    def __init__(self):
+        self._options = None  # Initialises in each thread
+
+    def get(self, key, default=None):
+        return self._options.get(key, default)
+
+
+local_context = ThreadContext()
+
+log = logging.getLogger(__name__)
+
+
+class TelluricContextError(Exception):
+    pass
+
+
+class TelluricContext(object):
+    """TelluricContext is a thread.local context manager for telluric parameters,
+    that can be used in telluric functionalities.
+
+    Example when applying a product on a raster for setting sensor_bands_info:
+        with TelluricContext(sensor_bands_info=bands_mapping):
+            return raster.apply('NDVI')
+
+    The arguments are saved in the `context` module under `local_contex` and to read them just use `local_context.get("argument_name"), for example:
+        
+        from telluric.context import local_context, TelluricContext
+        
+        with TelluricContex('a'=11):
+            local_context.get('a')  # => 11
+            local_context.get('b')  # => None
+            local_context.get('c', 44)  # => 44
+
+    TelluricContext can be nested and in that case the inner context overides the outer, for example:
+        from telluric.context import local_context, TelluricContext
+        
+        with TelluricContex('a'=11):
+            with TelluricContex('a'=33, 'b'='some_str'):
+                local_context.get('a')  # => 33
+                local_context.get('b')  # => some_str
+            local_context.get('a')  # => 11
+            local_context.get('b')  # => None
+    """
+
+    @classmethod
+    def default_options(cls):
+        """Default configuration options
+        Parameters
+        ----------
+        None
+        Returns
+        -------
+        dict
+        """
+        return {
+        }
+
+    def __init__(
+            self, **options):
+        """Create a new telluricContext settings.
+        Parameters
+        ----------
+        **options : optional
+            A mapping of key:value options, e.g.,
+            `CPL_DEBUG=True, CHECK_WITH_INVERT_PROJ=False`.
+
+        Returns
+        -------
+        TelluricContext
+        """
+        self.options = options.copy()
+        self.context_options = {}
+
+    @classmethod
+    def from_defaults(cls, *args, **kwargs):
+        """Create a context with default config options
+        Parameters
+        ----------
+        args : optional
+            Positional arguments for Env()
+        kwargs : optional
+            Keyword arguments for Env()
+        Returns
+        -------
+        TelluricContext
+        Notes
+        -----
+        The items in kwargs will be overlaid on the default values.
+        """
+        options = TelluricContext.default_options()
+        options.update(**kwargs)
+        return TelluricContext(*args, **options)
```

Comment:
What about using `self.__class__(*args, **options)` here and in all simillar places?

Reply:
can I do it on a classmethod?

Reply:
Did not notice that it is classmethod. Then even easier: `cls(*args, **options)`.